### ISSUE TITLE
ModuleNotFoundError: No module named 'deepspeed'

### ISSUE THREAD

**Issue by StoryHack** [neonbjb/tortoise-tts#556]
If i try to use read.py, or I try to enable deepspeed in do_tts.py, I get an error that says, "ModuleNotFoundError: No module named 'deepspeed'"

Here is the command I ran and the output

```
python tortoise/read.py --textfile 000001.txt --voice train_empire --preset fast 
Traceback (most recent call last):
  File "/home/bryce/Apps/tortoise-tts/tortoise/read.py", line 33, in <module>
    tts = TextToSpeech(models_dir=args.model_dir, use_deepspeed=args.use_deepspeed, kv_cache=args.kv_cache, half=args.half)
  File "/home/bryce/Apps/tortoise-tts/tortoise/api.py", line 231, in __init__
    self.autoregressive.post_init_gpt2_config(use_deepspeed=use_deepspeed, kv_cache=kv_cache, half=self.half)
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/TorToiSe-2.6.0-py3.9.egg/tortoise/models/autoregressive.py", line 374, in post_init_gpt2_config
    import deepspeed
ModuleNotFoundError: No module named 'deepspeed'
```

I tried to install deepspeed using pip, (pip install deepspeed), but that gave me the following error:

```
(tortoise) ... $ pip install deepspeed
Collecting deepspeed
  Downloading deepspeed-0.10.0.tar.gz (836 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 836.6/836.6 kB 6.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error
  
  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [8 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-f3llsto9/deepspeed_34b32072ae014276bf9e7cb62f021ffc/setup.py", line 83, in <module>
          cuda_major_ver, cuda_minor_ver = installed_cuda_version()
        File "/tmp/pip-install-f3llsto9/deepspeed_34b32072ae014276bf9e7cb62f021ffc/op_builder/builder.py", line 41, in installed_cuda_version
          assert cuda_home is not None, "CUDA_HOME does not exist, unable to compile CUDA op(s)"
      AssertionError: CUDA_HOME does not exist, unable to compile CUDA op(s)
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

```

And I'm not sure what to set CUDA_HOME to, or if that is a sign of something else. Any thoughts?

**Comment by marcelschliesser** [comment]
Hi @StoryHack, in the [docs](https://pypi.org/project/deepspeed/) of the **deepspeed package** is mentioned, that you need a [CUDA driver](https://developer.nvidia.com/cuda-downloads) installed.

Do you have it on your machine?

**Comment by StoryHack** [comment]
Is the CUDA driver not part of the pytorch-cuda package that is included in the installation instructions?

I just tried to install the CUDA toolkit seperately following the info on the link you provided, and then I was able to install deepspeed, but then when I tried to run read.py, I got page after page of warnings and error messages that whizzed by, ending with "RuntimeError: Error building extension 'transformer_inference'". Not sure what I did wrong, but I'll see if I can start over tomorrow.

**Comment by CreeperNinjaYT** [comment]
> Is the CUDA driver not part of the pytorch-cuda package that is included in the installation instructions?
> 
> I installed the CUDA toolkit following the info on the link you provided, and then I was able to install deepspeed, but then when I tried to run read.py, I got page after page of warnings and error messages that whizzed by, ending with "RuntimeError: Error building extension 'transformer_inference'". Not sure what I did wrong, but I'll see if I can start over tomorrow.

I have also done this and am getting the same error, both before and after installing cuda driver (i'm on linux BTW)

**Comment by manmay-nakhashi** [comment]
do export CUDA_HOME="path_to_cuda"

**Comment by StoryHack** [comment]
Ok, I removed the environment and started over, And I am still getting the error I mentioned . Here is the steps I took to install and attempt to get deepspeed working:
```
conda create --name tortoise python=3.9 numba inflect
conda activate tortoise
conda install pytorch torchvision torchaudio pytorch-cuda=11.7 -c pytorch -c nvidia
conda install -c "nvidia/label/cuda-11.7.1" cuda-toolkit
conda install transformers=4.29.2
pip install deepspeed
git clone https://github.com/neonbjb/tortoise-tts.git
cd tortoise-tts
python setup.py install
```
The only thing changed from the install instructions was adding the line to install cuda-toolkit (And forcing it to be the version that matches the pytorch-cuda stuff that is installed), and then use pip to install deepspeed. I tried one more time with the latest cuda-toolkit as well, with the same results as I will now describe.

If I run:

```
python tortoise/do_tts.py --text "I'm going to speak this" --voice random --preset fast 
```

It generated the 3 audio files, as expected. But deepspeed is off by default there. When I run...

```
python tortoise/do_tts.py --text "I'm going to speak this" --voice random --preset fast --use_deepspeed True
```

... there is apprently some kind of JIT compilation it is trying to perform. The output starts with:

```
[2023-08-09 11:18:04,072] [INFO] [real_accelerator.py:133:get_accelerator] Setting ds_accelerator to cuda (auto detect)
[2023-08-09 11:18:04,323] [INFO] [logging.py:96:log_dist] [Rank -1] DeepSpeed info: version=0.10.0, git-hash=unknown, git-branch=unknown
[2023-08-09 11:18:04,324] [WARNING] [config_utils.py:69:_process_deprecated_field] Config parameter mp_size is deprecated use tensor_parallel.tp_size instead
[2023-08-09 11:18:04,324] [INFO] [logging.py:96:log_dist] [Rank -1] quantize_bits = 8 mlp_extra_grouping = False, quantize_groups = 1
Using /home/bryce/.cache/torch_extensions/py39_cu117 as PyTorch extensions root...
Detected CUDA files, patching ldflags
Emitting ninja build file /home/bryce/.cache/torch_extensions/py39_cu117/transformer_inference/build.ninja...
Building extension module transformer_inference...
```

It then gives a bunch of output regarding the nvcc compilation. Each output line begins with the step [1/11], [2/11], etc. Each step appears to run fine until it gets to step 10. Here is the output from that point on:

```
[10/11] c++ -MMD -MF pt_binding.o.d -DTORCH_EXTENSION_NAME=transformer_inference -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1011\" -I/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/includes -I/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/includes -isystem /home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/include -isystem /home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/include/torch/csrc/api/include -isystem /home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/include/TH -isystem /home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/include/THC -isystem /home/bryce/miniconda3/envs/tortoise/include -isystem /home/bryce/miniconda3/envs/tortoise/include/python3.9 -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC -std=c++17 -O3 -std=c++17 -g -Wno-reorder -DBF16_AVAILABLE -c /home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp -o pt_binding.o 
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp: In instantiation of ‘std::vector<at::Tensor> ds_softmax_context(at::Tensor&, at::Tensor&, int, bool, bool, int, float, bool, bool, int, bool, unsigned int, unsigned int, at::Tensor&) [with T = float]’:
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1978:5:   required from here
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:536:50: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  536 |                                      {hidden_dim * InferenceContext::Instance().GetMaxTokenLength(),
      |                                       ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:536:50: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:537:41: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  537 |                                       k * InferenceContext::Instance().GetMaxTokenLength(),
      |                                       ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:537:41: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:545:38: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  545 |                          {hidden_dim * InferenceContext::Instance().GetMaxTokenLength(),
      |                           ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:545:38: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:546:29: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  546 |                           k * InferenceContext::Instance().GetMaxTokenLength(),
      |                           ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:546:29: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp: In instantiation of ‘std::vector<at::Tensor> ds_rms_mlp_gemm(at::Tensor&, at::Tensor&, at::Tensor&, at::Tensor&, at::Tensor&, float, at::Tensor&, at::Tensor&, bool, int, bool) [with T = float]’:
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1978:5:   required from here
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1571:72: warning: narrowing conversion of ‘(size_t)mlp_1_out_neurons’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
 1571 |         at::from_blob(intermediate_ptr, {input.size(0), input.size(1), mlp_1_out_neurons}, options);
      |                                                                        ^~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1571:72: warning: narrowing conversion of ‘mlp_1_out_neurons’ from ‘const size_t’ {aka ‘const long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp: In instantiation of ‘std::vector<at::Tensor> ds_softmax_context(at::Tensor&, at::Tensor&, int, bool, bool, int, float, bool, bool, int, bool, unsigned int, unsigned int, at::Tensor&) [with T = __half]’:
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1979:5:   required from here
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:536:50: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  536 |                                      {hidden_dim * InferenceContext::Instance().GetMaxTokenLength(),
      |                                       ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:536:50: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:537:41: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  537 |                                       k * InferenceContext::Instance().GetMaxTokenLength(),
      |                                       ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:537:41: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:545:38: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  545 |                          {hidden_dim * InferenceContext::Instance().GetMaxTokenLength(),
      |                           ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:545:38: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:546:29: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  546 |                           k * InferenceContext::Instance().GetMaxTokenLength(),
      |                           ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:546:29: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp: In instantiation of ‘std::vector<at::Tensor> ds_rms_mlp_gemm(at::Tensor&, at::Tensor&, at::Tensor&, at::Tensor&, at::Tensor&, float, at::Tensor&, at::Tensor&, bool, int, bool) [with T = __half]’:
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1979:5:   required from here
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1571:72: warning: narrowing conversion of ‘(size_t)mlp_1_out_neurons’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
 1571 |         at::from_blob(intermediate_ptr, {input.size(0), input.size(1), mlp_1_out_neurons}, options);
      |                                                                        ^~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1571:72: warning: narrowing conversion of ‘mlp_1_out_neurons’ from ‘const size_t’ {aka ‘const long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp: In instantiation of ‘std::vector<at::Tensor> ds_softmax_context(at::Tensor&, at::Tensor&, int, bool, bool, int, float, bool, bool, int, bool, unsigned int, unsigned int, at::Tensor&) [with T = __nv_bfloat16]’:
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1981:5:   required from here
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:536:50: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  536 |                                      {hidden_dim * InferenceContext::Instance().GetMaxTokenLength(),
      |                                       ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:536:50: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:537:41: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  537 |                                       k * InferenceContext::Instance().GetMaxTokenLength(),
      |                                       ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:537:41: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:545:38: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  545 |                          {hidden_dim * InferenceContext::Instance().GetMaxTokenLength(),
      |                           ~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:545:38: warning: narrowing conversion of ‘(((size_t)hidden_dim) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:546:29: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
  546 |                           k * InferenceContext::Instance().GetMaxTokenLength(),
      |                           ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:546:29: warning: narrowing conversion of ‘(((size_t)k) * (& InferenceContext::Instance())->InferenceContext::GetMaxTokenLength())’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp: In instantiation of ‘std::vector<at::Tensor> ds_rms_mlp_gemm(at::Tensor&, at::Tensor&, at::Tensor&, at::Tensor&, at::Tensor&, float, at::Tensor&, at::Tensor&, bool, int, bool) [with T = __nv_bfloat16]’:
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1981:5:   required from here
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1571:72: warning: narrowing conversion of ‘(size_t)mlp_1_out_neurons’ from ‘size_t’ {aka ‘long unsigned int’} to ‘long int’ [-Wnarrowing]
 1571 |         at::from_blob(intermediate_ptr, {input.size(0), input.size(1), mlp_1_out_neurons}, options);
      |                                                                        ^~~~~~~~~~~~~~~~~
/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/csrc/transformer/inference/csrc/pt_binding.cpp:1571:72: warning: narrowing conversion of ‘mlp_1_out_neurons’ from ‘const size_t’ {aka ‘const long unsigned int’} to ‘long int’ [-Wnarrowing]
[11/11] c++ pt_binding.o gelu.cuda.o relu.cuda.o layer_norm.cuda.o rms_norm.cuda.o softmax.cuda.o dequantize.cuda.o apply_rotary_pos_emb.cuda.o transform.cuda.o pointwise_ops.cuda.o -shared -lcurand -L/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/lib -lc10 -lc10_cuda -ltorch_cpu -ltorch_cuda -ltorch -ltorch_python -L/home/bryce/miniconda3/envs/tortoise/lib64 -lcudart -o transformer_inference.so
FAILED: transformer_inference.so 
c++ pt_binding.o gelu.cuda.o relu.cuda.o layer_norm.cuda.o rms_norm.cuda.o softmax.cuda.o dequantize.cuda.o apply_rotary_pos_emb.cuda.o transform.cuda.o pointwise_ops.cuda.o -shared -lcurand -L/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/lib -lc10 -lc10_cuda -ltorch_cpu -ltorch_cuda -ltorch -ltorch_python -L/home/bryce/miniconda3/envs/tortoise/lib64 -lcudart -o transformer_inference.so
/usr/bin/ld: cannot find -lcurand: No such file or directory
/usr/bin/ld: cannot find -lcudart: No such file or directory
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
Traceback (most recent call last):
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1893, in _run_ninja_build
    subprocess.run(
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/bryce/Apps/tortoise-tts/tortoise/do_tts.py", line 29, in <module>
    tts = TextToSpeech(models_dir=args.model_dir, use_deepspeed=args.use_deepspeed, kv_cache=args.kv_cache, half=args.half)
  File "/home/bryce/Apps/tortoise-tts/tortoise/api.py", line 231, in __init__
    self.autoregressive.post_init_gpt2_config(use_deepspeed=use_deepspeed, kv_cache=kv_cache, half=self.half)
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/TorToiSe-2.6.0-py3.9.egg/tortoise/models/autoregressive.py", line 375, in post_init_gpt2_config
    self.ds_engine = deepspeed.init_inference(model=self.inference_model,  
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/__init__.py", line 342, in init_inference
    engine = InferenceEngine(model, config=ds_inference_config)
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/inference/engine.py", line 192, in __init__
    self._apply_injection_policy(config)
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/inference/engine.py", line 426, in _apply_injection_policy
    replace_transformer_layer(client_module, self.module, checkpoint, config, self.config)
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/module_inject/replace_module.py", line 523, in replace_transformer_layer
    replaced_module = replace_module(model=model,
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/module_inject/replace_module.py", line 766, in replace_module
    replaced_module, _ = _replace_module(model, policy, state_dict=sd)
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/module_inject/replace_module.py", line 847, in _replace_module
    _, layer_id = _replace_module(child,
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/module_inject/replace_module.py", line 847, in _replace_module
    _, layer_id = _replace_module(child,
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/module_inject/replace_module.py", line 823, in _replace_module
    replaced_module = policies[child.__class__][0](child,
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/module_inject/replace_module.py", line 500, in replace_fn
    new_module = replace_with_policy(child,
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/module_inject/replace_module.py", line 348, in replace_with_policy
    _container.create_module()
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/module_inject/containers/gpt2.py", line 20, in create_module
    self.module = DeepSpeedGPTInference(_config, mp_group=self.mp_group)
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/model_implementations/transformers/ds_gpt.py", line 20, in __init__
    super().__init__(config, mp_group, quantize_scales, quantize_groups, merge_count, mlp_extra_grouping)
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/model_implementations/transformers/ds_transformer.py", line 58, in __init__
    inference_module = builder.load()
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/op_builder/builder.py", line 454, in load
    return self.jit_load(verbose)
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed/ops/op_builder/builder.py", line 497, in jit_load
    op_module = load(name=self.name,
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1284, in load
    return _jit_compile(
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1509, in _jit_compile
    _write_ninja_file_and_build_library(
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1624, in _write_ninja_file_and_build_library
    _run_ninja_build(
  File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch/utils/cpp_extension.py", line 1909, in _run_ninja_build
    raise RuntimeError(message) from e
RuntimeError: Error building extension 'transformer_inference'
```

It's the error I described in my above comment. My gut says it some kind of versioning/dependancy error, but I'm a little in over my head on this one, so I don't know which one to try.

The deepspeed library has a tool, ds_report, that can supposedly help track down issues when you know what it all means, which I don't. Here is the output from running ds_report:

```
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.0
 [WARNING]  using untested triton version (2.0.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/torch']
torch version .................... 2.0.1
deepspeed install path ........... ['/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/deepspeed']
deepspeed info ................... 0.10.0, unknown, unknown
torch cuda version ............... 11.7
torch hip version ................ None
nvcc version ..................... 11.7
deepspeed wheel compiled w. ...... torch 2.0, cuda 11.7
```

So it is trying to build that transformer_inference, as it is not installed, but it says that "op" is OKAY to build. I could try to clear up that async_io warning. Also I don't know if tortoise would work at all with an earlier torch version.

Any thoughts on where to go from here?

**Comment by Dharan2001** [comment]
> If i try to use read.py, or I try to enable deepspeed in do_tts.py, I get an error that says, "ModuleNotFoundError: No module named 'deepspeed'"
> 
> Here is the command I ran and the output
> 
> ```
> python tortoise/read.py --textfile 000001.txt --voice train_empire --preset fast 
> Traceback (most recent call last):
>   File "/home/bryce/Apps/tortoise-tts/tortoise/read.py", line 33, in <module>
>     tts = TextToSpeech(models_dir=args.model_dir, use_deepspeed=args.use_deepspeed, kv_cache=args.kv_cache, half=args.half)
>   File "/home/bryce/Apps/tortoise-tts/tortoise/api.py", line 231, in __init__
>     self.autoregressive.post_init_gpt2_config(use_deepspeed=use_deepspeed, kv_cache=kv_cache, half=self.half)
>   File "/home/bryce/miniconda3/envs/tortoise/lib/python3.9/site-packages/TorToiSe-2.6.0-py3.9.egg/tortoise/models/autoregressive.py", line 374, in post_init_gpt2_config
>     import deepspeed
> ModuleNotFoundError: No module named 'deepspeed'
> ```
> 
> I tried to install deepspeed using pip, (pip install deepspeed), but that gave me the following error:
> 
> ```
> (tortoise) ... $ pip install deepspeed
> Collecting deepspeed
>   Downloading deepspeed-0.10.0.tar.gz (836 kB)
>      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 836.6/836.6 kB 6.5 MB/s eta 0:00:00
>   Preparing metadata (setup.py) ... error
>   error: subprocess-exited-with-error
>   
>   × python setup.py egg_info did not run successfully.
>   │ exit code: 1
>   ╰─> [8 lines of output]
>       Traceback (most recent call last):
>         File "<string>", line 2, in <module>
>         File "<pip-setuptools-caller>", line 34, in <module>
>         File "/tmp/pip-install-f3llsto9/deepspeed_34b32072ae014276bf9e7cb62f021ffc/setup.py", line 83, in <module>
>           cuda_major_ver, cuda_minor_ver = installed_cuda_version()
>         File "/tmp/pip-install-f3llsto9/deepspeed_34b32072ae014276bf9e7cb62f021ffc/op_builder/builder.py", line 41, in installed_cuda_version
>           assert cuda_home is not None, "CUDA_HOME does not exist, unable to compile CUDA op(s)"
>       AssertionError: CUDA_HOME does not exist, unable to compile CUDA op(s)
>       [end of output]
>   
>   note: This error originates from a subprocess, and is likely not a problem with pip.
> error: metadata-generation-failed
> 
> × Encountered error while generating package metadata.
> ╰─> See above for output.
> 
> note: This is an issue with the package mentioned above, not pip.
> hint: See above for details.
> ```
> 
>

**Comment by Dharan2001** [comment]
i also have the same issue...have you got solution...if you got help me bro

**Comment by StoryHack** [comment]
No solution yet. Well, I can install deepspeed, (see the bit at the top of my big long comment) but it still doesn't work. I had to modify read.py so that the default is to not use deepspeed. I changed Line 27 to read:

```
parser.add_argument('--use_deepspeed', type=bool, help='Use deepspeed for speed bump.', default=False)
```

It will run fine, just without the deepspeed boost.

**Comment by CreeperNinjaYT** [comment]
> No solution yet. Well, I can install deepspeed, (see the bit at the top of my big long comment) but it still doesn't work. I had to modify read.py so that the default is to not use deepspeed. I changed Line 27 to read:
> 
> ```
> parser.add_argument('--use_deepspeed', type=bool, help='Use deepspeed for speed bump.', default=False)
> ```
> 
> It will run fine, just without the deepspeed boost.

Thank you for this, this is good enough for now : )

**Comment by Cartooninspector** [comment]
I use Tortoise TTS Oneclick 6.0 on windows.

Deepspeed only works well on windows on cu117 with all right torch versions.
You also have to add settings to start_windows.bat and type them every time you want to install something manually. Don't know if windows environment variables will work, didn't check it yet. But this works for me. I have a fully working tts one click webui with deepspeed, triton, bark, rvc, musicgen, etc.. You don't have to reïnstall conda or something. If you want to start all over. Just begin with a clean ttsone folder for install. You can remove conda env if you want, but it doesn't have to.

Install CUDA on windows. You can install as much as you want, because they just use the folder content. Don't forget to copy all cunn files to all dirs from versions you installed. You can use other versions for other apps. I installed Cuda 11.7; 11.8; 12.1

Be sure to have the miniconda3 installed with python 3.10. All scripts depend on python 3.10. I use. 23.11.0. Just like CUDA, you can install multiple versions on your windows. (3.10.x, 3.11.x, etc) Install 3.10 on your windows for sure.

So if you have CUDA, python and miniconda installed, you can go build:

Make sure you are in a dos prompt en start from here the batch files.
Edit start_windows.bat. Put on top:

@echo off
set DS_BUILD_OPS=0
set DS_BUILD_AIO=0 
set CUDA_VISIBLE_DEVICES=0,1
set FORCE_CUDA="1"

save and exit.

Just install start_windows.bat, after that update_windows.bat
When finished start:

(everything within the quotes)
"pip install torch==2.0.0+cu117 torchvision==0.15.1+cu117 torchaudio==2.0.1+cu117 -f https://download.pytorch.org/whl/torch_stable.html"

Deepspeed also has to be installed at the same environment. When you start tortoireTTS it doesn't activate one. 
Just open a normal cmd and go to your tts folder.

Now install Deepspeed:

cmd in admin mode:

@echo off
set DS_BUILD_OPS=0
set DS_BUILD_AIO=0 
set CUDA_VISIBLE_DEVICES=0,1
set FORCE_CUDA="1"

git clone  https://github.com/microsoft/DeepSpeed
cd deepspeed
build_win.bat
cd dist
pip install <build whl file.

go to tts folder and start start_windows.bat

You can also install triton in such way as deepspeed (don't forget the set commands before starting anything)
everything sould work now (i hope I didn't forget.
